### PR TITLE
Snowbridge: Synchronize from Snowfork repository

### PR DESCRIPTION
**Pull request by claravanstaden**
This PR includes the following 2 improvements:

## Ethereum Client

Author: @yrong 
### Original Upstream PRs
- https://github.com/Snowfork/polkadot-sdk/pull/123
- https://github.com/Snowfork/polkadot-sdk/pull/125

### Description
The Ethereum client syncs beacon headers as they are finalized, and imports every execution header. When a message is received, it is verified against the import execution header. This is unnecessary, since the execution header can be sent with the message as proof. The recent Deneb Ethereum upgrade made it easier to locate the relevant beacon header from an execution header, and so this improvement was made possible. This resolves a concern @svyatonik had in our initial Rococo PR: https://github.com/paritytech/polkadot-sdk/pull/2522#discussion_r1431270691

## Inbound Queue

Author: @yrong 
### Original Upstream PR
- https://github.com/Snowfork/polkadot-sdk/pull/118

### Description
When the AH sovereign account (who pays relayer rewards) is depleted, the inbound message will not fail. The relayer just will not receive rewards.

Both these changes were done by @yrong, many thanks. ❤️

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5584234

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5634461

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5640593

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5640755

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5640775

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5640773

File: cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/lib.rs [hunk 1]
```diff
@@ -643,7 +641,7 @@ parameter_types! {
 }
 
 parameter_types! {
-	pub const MaxExecutionHeadersToKeep: u32 = prod_or_fast!(8192 * 2, 1000);
+	pub const MaxExecutionHeadersToKeep: u32 = 8192 * 20;
```
Comment (yrong) [inline cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/lib.rs:644]: Seems not in use now.

Comment (yrong) [reply cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/lib.rs:644]: @claravanstaden I've added https://github.com/Snowfork/polkadot-sdk/pull/129 for that.

**Comment by claravanstaden** [comment]
Cleanup of storage will be done once the Multi-Block Migration feature has been configured on BridgeHub, since deleting Execution Headers in storage cannot be done in a single block.

Snowfork team @alistair-singh @vgeddes @yrong please review. :)

**Review by yrong — APPROVED** [review]

**Review by alistair-singh — APPROVED** [review]

File: bridges/snowbridge/pallets/ethereum-client/src/lib.rs [hunk 2]
```diff
@@ -191,25 +174,6 @@ pub mod pallet {
 	pub(super) type NextSyncCommittee<T: Config> =
 		StorageValue<_, SyncCommitteePrepared, ValueQuery>;
 
-	/// Latest imported execution header
```
Comment (svyatonik) [inline bridges/snowbridge/pallets/ethereum-client/src/lib.rs:194]: A migration for removal those values will be added later, right?

File: bridges/snowbridge/pallets/inbound-queue/src/lib.rs [hunk 3]
```diff
@@ -261,11 +262,19 @@ pub mod pallet {
...
 			let sovereign_account = sibling_sovereign_account::<T>(channel.para_id);
 			let delivery_cost = Self::calculate_delivery_cost(message.encode().len() as u32);
-			T::Token::transfer(&sovereign_account, &who, delivery_cost, Preservation::Preserve)?;
+			let amount = T::Token::reducible_balance(
+				&sovereign_account,
+				Preservation::Preserve,
+				Fortitude::Polite,
+			)
+			.min(delivery_cost);
+			if amount > BalanceOf::<T>::zero() {
```
Comment (svyatonik) [inline bridges/snowbridge/pallets/inbound-queue/src/lib.rs:275]: Could be simplified to:
```suggestion
			if !amount.is_zero() {
```

but feel free to ignore, since it'll require another PR in another repo :)

**Review by svyatonik — APPROVED** [review]
Have been mostly checking direct changes to pallet code - all looks great, thank you!

File: bridges/snowbridge/pallets/ethereum-client/src/lib.rs (hunk 2 above)
Comment (claravanstaden) [reply bridges/snowbridge/pallets/ethereum-client/src/lib.rs:194]: Yes, definitely. As soon as Multi-block Migrations are ready on BH.

File: bridges/snowbridge/pallets/ethereum-client/src/impls.rs [hunk 4]
```diff
@@ -53,18 +36,11 @@ impl<T: Config> Verifier for Pallet<T> {
 		if !receipt.contains_log(&event_log) {
 			log::error!(
 				target: "ethereum-client",
-				"💫 Event log not found in receipt for transaction at index {} in block {}",
-				proof.tx_index, proof.block_hash,
+				"💫 Event log not found in receipt for transaction",
```
Comment (bkontur) [inline bridges/snowbridge/pallets/ethereum-client/src/impls.rs:39]: Why did you remove log params? I think this log became pretty non-relevant and useless, because you also remove input log with block_hash `Verifying message with block hash`.
How will you investigate `Event log not found in receipt for transaction` without knowing block_hash or tx?

File: bridges/snowbridge/pallets/ethereum-client/src/lib.rs (hunk 2 above)
Comment (bkontur) [reply bridges/snowbridge/pallets/ethereum-client/src/lib.rs:194]: maybe this could be useful: https://github.com/paritytech/polkadot-sdk/pull/3828

**Review by bkontur — APPROVED** [review]

File: prdoc/pr_3761.prdoc [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# Schema: Polkadot SDK PRDoc Schema (prdoc) v1.0.0
+# See doc at https://raw.githubusercontent.com/paritytech/polkadot-sdk/master/prdoc/schema_user.json
+
+title: Snowbridge Ethereum client improvements and prevent relayer rewards from message processing to fail
```
Comment (bkontur) [inline prdoc/pr_3761.prdoc:4]: @claravanstaden iirc, this title needs to match PR title, so please better change one of them. @EgorPopelyaev right?

File: bridges/snowbridge/pallets/ethereum-client/src/impls.rs (hunk 4 above)
Comment (claravanstaden) [reply bridges/snowbridge/pallets/ethereum-client/src/impls.rs:39]: The reasoning behind this is on the original PR, over here: https://github.com/Snowfork/polkadot-sdk/pull/125#discussion_r1529183021 

A side-note, we don't generally have access to the parachain logs either, so we try and solve issues without having to look at the logs.

File: prdoc/pr_3761.prdoc (hunk 5 above)
Comment (acatangiu) [reply prdoc/pr_3761.prdoc:4]: No, it doesn't need to match. E.g. I just merged https://github.com/paritytech/polkadot-sdk/pull/3927 with prdoc with slightly different title than PR.

Reason it's failing is missing `bump` qualifier. See e.g. https://github.com/acatangiu/polkadot-sdk/blob/67e322206ea9d1990baae560c73a46e72dff5b2d/prdoc/pr_3930.prdoc#L10

File: bridges/snowbridge/pallets/ethereum-client/src/impls.rs (hunk 4 above)
Comment (bkontur) [reply bridges/snowbridge/pallets/ethereum-client/src/impls.rs:39]: ok, got it, and what about adding there some "identifier" from `proof.execution_proof` or `proof.execution_proof.execution_header`, does it make sense?

File: prdoc/pr_3761.prdoc (hunk 5 above)
Comment (claravanstaden) [reply prdoc/pr_3761.prdoc]: Updated prdoc over here: https://github.com/paritytech/polkadot-sdk/pull/3761/commits/ef97c0bbc3d6bf5607fb2112bc6d5d4ad4d11c25 I marked the Snowbridge crate version bumps as major because this is a major, and breaking change (from an interface perspective).

File: bridges/snowbridge/pallets/ethereum-client/src/impls.rs (hunk 4 above)
Comment (claravanstaden) [reply bridges/snowbridge/pallets/ethereum-client/src/impls.rs:39]: I don't know if it would be useful. We have the same info in the relayer and that is more accessible to us.

File: prdoc/pr_3761.prdoc (hunk 5 above)
Comment (claravanstaden) [reply prdoc/pr_3761.prdoc]: Fixed :)

File: bridges/snowbridge/pallets/ethereum-client/src/lib.rs (hunk 2 above)
Comment (claravanstaden) [reply bridges/snowbridge/pallets/ethereum-client/src/lib.rs:194]: Looks great! Will that feature be OK even if all the storage items cannot be deleted in one block?
